ipld/resolver/tests/smoke.rs

```diff
@@ -0,0 +1,250 @@
+        Self {
+            size,
+            rng: rand::rngs::StdRng::seed_from_u64(seed),
+            services: Default::default(),
+            agents: Default::default(),
+        }
+    }
+
+    /// Add a node with randomized address, optionally bootstrapping from an existing node.
+    fn add_node(&mut self, bootstrap: Option<usize>) {
+        let bootstrap_addr = bootstrap.map(|i| {
+            let config = &self.agents[i].config;
+            let peer_id = config.network.local_peer_id();
+            let mut addr = config.connection.listen_addr.clone();
+            addr.push(Protocol::P2p(peer_id.into()));
+            addr
+        });
+        let config = make_config(&mut self.rng, self.size, bootstrap_addr);
+        let (service, client, store) = make_service(config.clone());
+        self.services.push(service);
+        self.agents.push(Agent {
+            config,
+            client,
+            store,
+        });
+    }
+
+    /// Start running all services
+    fn run(self) -> Cluster {
+        for service in self.services {
+            tokio::task::spawn(async move { service.run().await.expect("error running service") });
+        }
+        Cluster {
+            agents: self.agents,
+        }
+    }
+}
+
+/// Start a cluster of agents from a single bootstrap node,
+/// make available some content on one agent and resolve it from another.
+#[tokio::test]
+async fn single_bootstrap_single_provider_resolve_one() {
+    let _ = env_logger::builder().is_test(true).try_init();
+    //env_logger::init();
+
+    // Choose agents.
+    let cluster_size = 3;
+    let bootstrap_idx = 0;
+    let provider_idx = 1;
+    let resolver_idx = 2;
+
+    // TODO: Get the seed from QuickCheck
+    let mut builder = ClusterBuilder::new(cluster_size, 123456u64);
+
+    // Build a cluster of nodes.
+    for i in 0..builder.size {
+        builder.add_node(if i == 0 { None } else { Some(bootstrap_idx) });
+    }
+
+    // Start the swarms.
+    let mut cluster = builder.run();
+
+    // Insert a CID of a complex recursive data structure.
+    let cid = insert_test_data(&mut cluster.agents[provider_idx]).expect("failed to insert data");
+
+    // Sanity check that we can read the data back.
+    check_test_data(&mut cluster.agents[provider_idx], &cid).expect("failed to read back the data");
+
+    // Wait a little for the cluster to connect.
+    // TODO: Wait on some condition instead of sleep.
+    tokio::time::sleep(Duration::from_secs(1)).await;
+
+    // Announce the support of some subnet.
+    let subnet_id = make_subnet_id(1001);
+
+    cluster.agents[provider_idx]
+        .client
+        .add_provided_subnet(subnet_id.clone())
+        .expect("failed to add provided subnet");
+
+    // Wait a little for the gossip to spread and peer lookups to happen, then another round of gossip.
+    // TODO: Wait on some condition instead of sleep.
+    tokio::time::sleep(Duration::from_secs(2)).await;
```

Comment:
This have been shown to be a recurring issue when testing gossipsub-related protocols. >1s sync delay (which is the default heartbeat time) is the easy one, but IIRC we can probably listen or poll the libp2p host to see if it is bootrstapped. I don't know if this can be done easily in Rust, I did this once in Go.

(We can do something similar with the bootstrap delay above)

Reply:
Thank you for the tips. I can also add more events to show when things have progressed, and I have seen examples in the libp2p tests of a combination of a polling function and a timeout.